[WIP][DARWIN] libgsm: 1.0.13 -> 1.0.14 & improve cross platform support

codyopel opened the pull request:
Default Makefile for reference
https://gist.github.com/codyopel/747154cc34874dcb6e4d
dylib patch example (this is a patch for an already patched version)
https://searchcode.com/codesearch/view/37251938/

@spwhitt
Incase you can offer any help.
I originally had to had a few hacks to get this to build a shared library correctly for ELF platforms.  I started adding workarounds for Darwin but I have no clue if they work.  Basically the Makefile needs to be patched to build a dylib instead of building statically preferably.  I left the duplication of arguments between the ELF and Darwin hacks for now in case they need to be modified for Darwin.

If all else fails we can remove the darwin dylib hack and just build is staticly which it will do by default, then we only have to make sure it uses clang.

spwhitt commented on pkgs/development/libraries/gsm/default.nix:
```diff
@@ -1,41 +1,51 @@
+    ++ optional isDarwin ''
+      sed -e 's,-c -O2 -DNeedFunctionPrototypes=1,-c -O2 -fPIC -DNeedFunctionPrototypes=1,' -i Makefile
+      sed -e 's,libgsm.a,libgsm.dylib,' -i Makefile
+      sed -e 's/$(AR) $(ARFLAGS) $(LIBGSM) $(GSM_OBJECTS)/$(LD) -o $(LIBGSM) -dynamiclib -install_name $(LIBGSM) $(GSM_OBJECTS) -lc/' -i Makefile
```
Change the install name to `$(INSTALL_ROOT)/$(LIBGSM)`. Don't forget to change the set the separator character to something other than /. Using an absolute path is best for dylibs in nix, it helps avoid a ton of bugs.

codyopel replied:
I though so, but wasn't sure if the flag also need the install path

spwhitt commented on pkgs/development/libraries/gsm/default.nix:
```diff
@@ -1,41 +1,51 @@
+    ''
+    # Build dylib on Darwin
+    ++ optional isDarwin ''
+      sed -e 's,-c -O2 -DNeedFunctionPrototypes=1,-c -O2 -fPIC -DNeedFunctionPrototypes=1,' -i Makefile
```
You don't need this, NIX_CFLAGS_COMPILE does the same thing.

codyopel replied:
There was a reason I did this, but I'll look into it, it's probably useless doing it this way.

spwhitt commented on pkgs/development/libraries/gsm/default.nix:
```diff
@@ -1,41 +1,51 @@
+    ''
+  ] # Build ELF shared library instead of static
+    ++ optional (!isCygwin && !isDarwin) ''
+      sed -e 's,-c -O2 -DNeedFunctionPrototypes=1,-c -O2 -fPIC -DNeedFunctionPrototypes=1,' -i Makefile
```
Don't need this for the same NIX_CFLAGS_COMPILE reason.

spwhitt commented on pkgs/development/libraries/gsm/default.nix:
```diff
@@ -1,41 +1,51 @@
+    # Fix prefix & include directories
+    ''
+      sed -e 's,$(GSM_INSTALL_ROOT)/inc,$(GSM_INSTALL_ROOT)/include/gsm,' -i Makefile
+      makeFlags="$makeFlags INSTALL_ROOT=$out"
```
You could set this in the makeFlags array if you prefer `"INSTALL_ROOT=$(out)"`

codyopel replied:
That was orginally a work around and I just haven't switched it back

spwhitt commented on pkgs/development/libraries/gsm/default.nix:
```diff
@@ -1,41 +1,51 @@
+    '';
 
   makeFlags = [
     ''SHELL=${stdenv.shell}''
```
Weird place to use this type of quote.

spwhitt commented:
Only the install name comment really matters, the rest is cosmetic. With the install name fix it seems to work for me. I'll test again after the update.

codyopel commented:
@spwhitt 
I added the lib install directory.  It should install to $out/lib/libgsm.dylib now.

The cflags aren't directly recognized by gsm's broken Makefile when using ```NIX_CFLAGS_COMPILE```, I changed around the sed hack to use it so that it is a little cleaner, unless you know of a cleaner way that works.

spwhitt commented on pkgs/development/libraries/gsm/default.nix:
```diff
@@ -1,41 +1,53 @@
+      sed -e 's/$(AR) $(ARFLAGS) $(LIBGSM) $(GSM_OBJECTS)/$(LD) -shared -Wl,-soname,libgsm.so -o $(LIBGSM) $(GSM_OBJECTS) -lc/' -i Makefile
+    '' ] ++ [ ''
+      # Hack to add cflags for shared libraries
+      sed -e 's,-c -O2 -DNeedFunctionPrototypes=1,-c -O2 -DNeedFunctionPrototypes=1 ${NIX_CFLAGS_COMPILE},' -i Makefile
```
I still don't think this is needed. NIX_CFLAGS_COMPILE isn't supposed to be recognized by GSM's makefile. It inserts the flags into the cc-wrapper.

codyopel replied:
Yeah, you're right, I didn't have time to play with it last night and test it.  Only change is that it has to be optional, but I added that.

spwhitt commented on pkgs/development/libraries/gsm/default.nix:
```diff
@@ -1,41 +1,53 @@
+      # Hack to add cflags for shared libraries
+      sed -e 's,-c -O2 -DNeedFunctionPrototypes=1,-c -O2 -DNeedFunctionPrototypes=1 ${NIX_CFLAGS_COMPILE},' -i Makefile
+      # Remove line that is unused when building shared libraries
+      sed -e 's,$(RANLIB) $(LIBGSM),,' -i Makefile
```
This is only getting applied for Linux, but is needed for Darwin as well. I think you didn't realize the `else` operator takes precedence over `++`

codyopel replied:
Yeah, normally I would catch a silly mistake like that, but since I can't test on darwin I didn't notice.  Also moving the closing parenthese would fix it too, but I like the optionalString method so I changed it around a bit to use that.

spwhitt commented:
For reference, here's how I chose to fix it when testing locally: https://github.com/spwhitt/nixpkgs/tree/codyopel-gsm-cross

codyopel committed:
libgsm: 1.0.13 -> 1.0.14 & improve cross platform support

spwhitt commented:
:+1: 

Works on Darwin. LGTM.

vcunat commented:
Well, I see libgsm.so having ~48kB on master. Is it even worth linking that dynamically? We don't save any rebuilds and it doesn't seem used in so many packages anyway.
